New function to copy MCInfo to output file outside the main loop for selected events

jmalbos opened the pull request:
This PR introduces a new function (`copy_mc_info`) in mcinfo_io to copy the MCInfo from the input files to the output file for a list of events.

jmalbos committed:
Add new function mcinfo_io.copy_mc_info

jmalbos committed:
Modify Penthesilea to use function copy_mc_info

jmalbos committed:
Remove MCInfo from data yielded by cities' sources

Three file readers (wf_from_files, pmap_from_files and
hits_and_kdst_from_files) yield the MC info bank as part of
the data they read from file. Remove that data from the yielded
dictionary since the MCInfo is now going to be copied to the
output file from outside the city's main loop using the function
components.copy_mc_info.

gonzaponte commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -3,6 +3,9 @@
 
+from .. cities.components import get_run_number
+from .. cities.components import get_mc_info_safe
+
```
by design, `components` should only be used in the cities. Those functions belong somewhere else.

jmalbos replied:
Not sure what am I suppose to do then; notice that those are pre-existing functions. Should I move them to` mcinfo_io`?

gonzaponte replied:
Well, as discussed below, you will not need them, you will need `get_mc_info`, which is in reco.tbl_function. I think that function should also be moved from there to mcinfo_io, but maybe that's for another PR.

gonzaponte commented on invisible_cities/cities/components.py:
```diff
@@ -277,7 +277,7 @@ def pmap_from_files(paths):
             for evtinfo in event_info:
                 event_number, timestamp = evtinfo.fetch_all_fields()
-                yield dict(pmap=pmaps[event_number], mc=mc_info,
+                yield dict(pmap=pmaps[event_number], #mc=mc_info,
```
Remove instead of commenting out.

gonzaponte commented on invisible_cities/cities/components.py:
```diff
@@ -267,7 +267,7 @@ def pmap_from_files(paths):
                 run_number  = get_run_number(h5in)
                 event_info  = get_event_info(h5in)
-                mc_info     = get_mc_info_safe(h5in, run_number)
+                #mc_info     = get_mc_info_safe(h5in, run_number)
```
Remove instead of commenting out.

gonzaponte commented on invisible_cities/cities/penthesilea.py:
```diff
@@ -85,28 +86,32 @@ def penthesilea(files_in, file_out, compression, event_range, print_mod, detecto
+                                    df.fork((build_hits           , write_hits           ),
+                                            (build_pointlike_event, write_pointlike_event),
+                                                                    events_selected       ,
+                                                                    write_event_info     )),
```
maybe change the order so it doesn't break the symmetry between the writes?

gonzaponte commented on invisible_cities/cities/penthesilea.py:
```diff
@@ -85,28 +86,32 @@ def penthesilea(files_in, file_out, compression, event_range, print_mod, detecto
-                                  events_out = event_count_out.future,
-                                  selection  = pmap_select    .future))
+        event_ids = []
+        events_selected = df.sink(event_ids.append, args=("event_number"))
```
Mmm. This is **not** wrong, but we tend to use futures for this kind of things. My approach would be to use the `reduce` component, but I can't think of a good reason not to take this approach. The only (very minor thing) is that using something named `result.events_selected` (or similar) would be slightly more indicative of what is happening, but it's not a big deal at all. I kinda like the fact that you can specify the argument to take :+1:

gonzaponte replied:
The parenthesis around `"event_number" can be removed.

jmalbos replied:
The main reason why I didn't use `reduce` in the end is that it forced me to write something like this:
`add_to_list = df.reduce(lambda l,x: l+[x], initial=[])`
It works, but it looks a bit ugly. I can change it though; it isn't particularly important.

gonzaponte replied:
No, I like this approach very much!

jacg replied:
TL:DR the `sink` approach is fine, but just for fun and education ...

> `add_to_list = df.reduce(lambda l,x: l+[x], initial=[])`

... is `O(N)` per insertion (where `N` is the length of the stream) and hence `O(N^2)` in total. You can *almost* do it in `amortized O(1)` per insertion, `O(N)` total with `add_to_list = df.reduce(list.append, initial=[])` except that `list.append` returns `None` rather than the list itself, which is a Python wart, as far as I am concerned, and throws a spanner in the works here.

But you could hack around it with `df.reduce(lambda l,x: l.append(x) or l, initial=[])`, which is a bit ugly, so, you could provide

````python
def append(list_, element):
    list_.append(element)
    return list_
````
which would allow you to say
````python
df.reduce(append, initial=[])
````
cleanly, and keep using the futures pattern that we are used to seeing, and thus address @gonzaponte's comment that

> using something named `result.events_selected` (or similar) would be slightly more indicative of what is happening

gonzaponte replied:
>  except that list.append returns None rather than the list itself

For that, we had thought of using `np.append` which does return the object, at the expense of making a copy of the array every time.

jacg replied:
Which makes the time-cost of the whole process quadratic rather than linear in the length of the source.

gonzaponte replied:
Yes, of course. I wasn't suggesting that it was a good choice, but rather an attempt to leave some trace that this was considered :)

jacg replied:
Yup, good.

Just one more reminder: it's great to theorize, but it's also good to back the theory up with measurement. For which `timeit` (and the way it is embedded in IPython) can be very useful:

````
In [1]: import numpy as np                                                                                                                                                                         

In [2]: from functools import reduce                                                                                                                                                               

In [3]: def a(N): 
   ...:     return reduce(np.append, range(N), np.array([])) 
   ...:

In [4]: def b(N): 
   ...:     return reduce(lambda l,e: l.append(e) or l, range(N), []) 
   ...:

In [5]: assert (a(10**3) == b(10**3)).all()

In [6]: for fn in (a,b): 
   ...:     for p in range(3,6): 
   ...:         %timeit fn(10**p) 
   ...:     print('='*60) 
   ...:

11.9 ms ± 74.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
143 ms ± 1.67 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
4.97 s ± 41.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
============================================================
211 µs ± 12.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
2.12 ms ± 32.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
21.4 ms ± 155 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
============================================================

In [7]: np.array([143, 4970]) / 11.9
Out[7]: array([ 12.01680672, 417.64705882])

In [8]: np.array([2120, 21400]) / 211
Out[8]: array([  10.04739336, 101.42180095])

````

In summary, the `lambda l,e: l.append(e) or l` approach is significantly faster for nontrivial input size: (211, 2120, 21400) vs (11900, 143000, 4970000); and grows more slowly with increasing size: (1, 10, 111) vs (1, 12, 418).

gonzaponte commented on invisible_cities/cities/penthesilea.py:
```diff
@@ -85,28 +86,32 @@ def penthesilea(files_in, file_out, compression, event_range, print_mod, detecto
+                                    events_out = event_count_out.future,
+                                    selection  = pmap_select    .future))
+
+        copy_mc_info(files_in, h5out, event_ids)
```
I think it is safer and more verbose for someone trying to understand the code to make `copy_mc_info` run only when the run number is `<= 0` and not let it figure it out from the run number in the file

gonzaponte replied:
Another side effect of calling `copy_mc_info` always is that the MC tables are created even for non-MC files!

jmalbos replied:
OK!

jmalbos replied:
Done.

gonzaponte commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -168,6 +172,37 @@ def __call__(self, mctables: (tb.Table, tb.Table, tb.Table, tb.Table),
+        try:
+            with tb.open_file(f, "r") as h5in:
+                run_number = get_run_number(h5in)
+                mc_info    = get_mc_info_safe(h5in, run_number)
```
Related to what I said in another comment. Do not let it figure it out. This is a base function and I think it is safer and better to make it work only for the case it is meant to be used. In my opinion, it should complain if there is not a MC table in the input file.

gonzaponte replied:
Apparently, this comment was confusing. I meant that `copy_mc_info` should not assume anything about whether the sign of run number means something or not. I would like `copy_mc_info` to assume whatever files come in to be MC files and have the relevant MC tables. If something goes wrong it should crash (there should be an explicit check to make the crash more obvious). Let the user of this function (aka the cities or the components) handle that situation (that is, calling `copy_mc_info` only for MC files).

jmalbos replied:
Done.

gonzaponte commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -168,6 +172,37 @@ def __call__(self, mctables: (tb.Table, tb.Table, tb.Table, tb.Table),
+                    except IndexError:
+                        # The event id was not found in the current file
+                        continue
+        except:
```
catch specific exceptions and, in case they are too generic, put some comment about what you are catching.

gonzaponte commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -168,6 +172,37 @@ def __call__(self, mctables: (tb.Table, tb.Table, tb.Table, tb.Table),
+                        event_list.remove(n)
+                    except IndexError:
+                        # The event id was not found in the current file
+                        continue
```
A few things:
- There is nothing to ensure that `event_list` is empty at the end of all loops... I wonder if we can make sure of that.
- Maybe we can get rid of this `try/except` clause by using `set`s. If you take the event numbers of the input file and make the intersection with `event_list` you would only run over the event numbers in that file without having to check them all.
- I think we don't have any protection against repeated event numbers in the writer, but it's something to think about (not for this PR, though)

mmkekic replied:
> Maybe we can get rid of this try/except clause by using sets. If you take the event numbers of the input file and make the intersection with event_list you would only run over the event numbers in that file without having to check them all.

This function will be modified in #693, so I dont think we need to fine tune it too much. However this implementation has 2 issues:

1. it sneakily modifies event_list
2. in case event numbers are the same across different input files just the first occurrence (file, event number) will be written.  I am not sure how big of an issue it is cause we do not have any protection anywhere against repeated event numbers in our readers and writers ...

jmalbos replied:
I can fix somewhat the _sneaky_ behavior of the function. However, there were several things about `event_list` I wasn't sure I could assume; namely:
- Do event ids come in order?
- Is that order correlated with the `files_in` order?
- Can event ids repeat?
I understand that the answer to the first two is 'no', and that we really don't know the answer of the third one, correct?

gonzaponte replied:
> Do event ids come in order?

Not necessarily

> Is that order correlated with the `files_in` order?

They will keep the order in which they are written within the same file, and the order of the input files is kept. That is if you have `file_1` with events `[2, 6, 4]` and `file_2` with events `[1, 3, 5]`
you will receive `[2, 6, 4, 1, 3, 5]`

> Can event ids repeat?

They can, in theory. It shouldn't happen, but there is nothing preventing it. For instance if you run over two different MC productions at the same time, the event numbers will repeat unless you take care of it explicitly. It can also happen if someone doesn't take care of organizing the event number in a `nexus` run that is split into many jobs. So far we haven't had issues, but I think it can happen in #641.

paolafer replied:
So far, we have assumed that event IDs don't repeat, because we were using dictionaries in the event data model, correct me if I'm wrong. I'm not sure this is the case anymore, so we may want to discuss this point. Not to repeat event IDs is particularly cumbersome when we do event mixing in the background model simulations, so, if we can drop this requirement, it may be a simplification.

mmkekic replied:
Now that i think of it the safest would be to have a dictionary `{file_in : event_list} `, that way there can not be confusion. However we will not be able to read this file since as @paolafer  said we do expect to have unique IDs at the reader level...

gonzaponte replied:
How about making this function for a single file and expand it for many in components? This would make the utility available in general (which is probably useful) while keeping these intricacies away from the base functions. I don't think we need to deal with repeated events or the ordering here. My feeling is that we should assume that order doesn't matter and that there shouldn't be repeated events for practical purposes. In case, we can add a tool that ensures that somewhere.

jmalbos replied:
> How about making this function for a single file and expand it for many in components? This would make the utility available in general (which is probably useful) while keeping these intricacies away from the base functions. I don't think we need to deal with repeated events or the ordering here. My feeling is that we should assume that order doesn't matter and that there shouldn't be repeated events for practical purposes. In case, we can add a tool that ensures that somewhere.

But how do I correlate files with selected events then? The main loop in push is reading all input files together.

gonzaponte replied:
I propose to do the following:

In `mcinfo_io.copy_mc_info`
- Make this function take `h5in`,`h5out` and `which_events` (unofficial names)
- If `which_events` is `None` (probably default), read the event numbers from the input file and assign it to `which_events`. This is a detail that doesn't affect the current usage, but that it would be good to have.
- Assume that each event in `which_events` is stored in the file.
- For each event in `which_events`, copy it.

In `components.copy_mc_info` (unofficial name):
- take `filenames_in`, `h5out` and `all_filtered_events`
- for each filename
  - open the file and call it `h5in` (unofficial name)
  - read the event numbers in that file and make the intersection with `all_filtered_events`. Let's call that `filtered_events_in_file`
  - call `mc_info_io.copy_mc_info(h5in, h5out, filtered_events_in_file)`

With this:
- Duplicated event numbers in different files are written
- The order of the files is irrelevant
- The order of the events within a file matters (as it has mattered so far), but hopefully this will be addressed in the future

jmalbos replied:
Sounds good, @gonzaponte.

mmkekic replied:
Also dont forget to change get_mc_info_safe to get_mc_info directly!

gonzaponte reviewed (commented):
We need some tests for the new function :)

jacg commented on invisible_cities/cities/penthesilea.py:
```diff
@@ -83,6 +82,14 @@ def penthesilea(files_in, file_out, compression, event_range, print_mod, detecto
+        l.append(e)
+        return l
+
+    add_to_list = df.reduce(append, initial=[])
```
I'd be tempted to

1. rename `add_to_list` to `collect_into_list`
2. contribute it to `dataflow` (If you do that, then rename `append` to `_append` when you put it into `dataflow`, or just use the `lambda` version; and, of course, add a documenting test.)

jmalbos committed:
Add function dataflow.collect_into_list and tests

– Add new function dataflow.collect_into_list that stores an argument
  in a list. The list can be retrieved from a future associated
  to a spy branch in a namedtuple.

- Add three tests of the new function: they test the use when no
  argument is provided, when one argument is provided and when
  multiple arguments are provided.

jmalbos committed:
Add new function 'components.collect'

The new function returns a future/sink pair for collecting streams
into a list.

jacg commented on invisible_cities/cities/components.py:
```diff
@@ -152,6 +152,17 @@ def print_every_loop(target):
+def collect():
+    """
+    Collect into a list an argument (selector string) passed through a pipe.
+    The list can be retrieved from a future.
```
I wouldn't mention the selector string, as it's completely orthogonal: `collect` collects whatever comes down the pipe, which might or might not have been selected with a selector string.

Saying that it can be retrieved from a future also seems a bit odd, because this is absolutely standard behaviour for any sink that produces a value. So, either don't mention the future at all, or describe fully how it returns a a future and a sink pair.

Also, https://www.python.org/dev/peps/pep-0257/#id17

OTOH, I wouldn't worry too much about all this, as it will be documented properly in the stand-alone dataflow package. So you might as well leave it as it is, for the purposes of this PR.

jmalbos committed:
Modify Diomira to use function copy_mc_info

jmalbos committed:
Modify Irene to use function copy_mc_info

jmalbos committed:
Modify Isidora to use function copy_mc_info

jmalbos committed:
Modify Esmeralda to use function copy_mc_info

jmalbos committed:
Add tests for components.collect and .copy_mc_info

- Add test of components.collect that evaluates that a simple data
  source consisting of a list of numbers is returned as a future.

- Add test of components.copy_mc_info that checks that the function
  can handle files without MC info.

– Add test of components.copy_mc_info that checks that the function
  cannot handle (i.e. the test fails) input files with repeated event
  numbers.

- Remove the check for MC info from a test in
  test_hits_and_kdst_from_files since it is no longer yielded by the
  hits_and_kdst_from_files file reader.

jmalbos committed:
Add new function components.copy_mc_info

Copy from a number of input files to an output file the MC info
of certain events identified by their event number.

jmalbos committed:
Add tests of mcinfo_io.copy_mc_info

- Add a test that checks whether the entire MC info of an input file
  is copied to the output file if no list of event numbers is provided.

- Add a test that evaluates the correct behavior of the function
  when a list of events to be copied is provided. That is, the
  output file has to contain only the MC info of the events specified
  in the list.

- Add a test checking that an assertion is raised when an event
  not included in the input file is specified in the list of events
  to be copied.

gonzaponte commented on invisible_cities/cities/components.py:
```diff
@@ -152,6 +154,39 @@ def print_every_loop(target):
+    return fl.reduce(append, initial=[])()
+
+
+def copy_mc_info(files_in : List[str], h5out: tb.File, event_numbers: List[int]):
```
return type hint `-> None`

gonzaponte commented on invisible_cities/cities/components.py:
```diff
@@ -152,6 +154,39 @@ def print_every_loop(target):
+                event_numbers_in_file = h5in.root.MC.extents.cols.evt_number[:]
+                event_numbers_to_copy = set(event_numbers_in_file).intersection(event_numbers)
+                mcinfo_io.copy_mc_info(h5in, writer, list(event_numbers_to_copy))
+            except tb.exceptions.NoSuchNodeError:
```
I don't think this exception should pass silently. If there is no data, it should complain.

mmkekic replied:
Maybe throwing a warning? Notice that we want to be able to run the cities on empty files

gonzaponte replied:
Ah, if that is the reason I guess it makes sense. But maybe we can introduce a function to detect empty files and use that. With this implementation we cannot distinguish between that and a file that has something wrong.

jmalbos replied:
> Ah, if that is the reason I guess it makes sense. But maybe we can introduce a function to detect empty files and use that. With this implementation we cannot distinguish between that and a file that has something wrong.

That is the reason, yes.

gonzaponte replied:
@mmkekic., do you know whether an empty file contains the `runInfo` table with zero entries? This could be the key to detect empty files vs "corrupt" files.

mmkekic replied:
They do! So maybe thats the first thing we can check per file_in, if its empty do nothing, if its not and the reading doesnt work raise an error

jmalbos replied:
The test `cities_test.test_city_empty_input_file` uses an empty input file that contains no tables.

gonzaponte replied:
I can't confirm, but I would say that this is not realistic. It depends on the decoder, which is not in this repo, but I think these tables are always there, even if empty. In any case, a function that checks whether a file is empty or not could check both whether they exist and, if so, check whether they contain any entries.

jmalbos replied:
Sure, but shouldn't that be a task for a different PR?

gonzaponte replied:
I would prefer that, but that would delay this one, right? I am open to introducing it here, if necessary. @mmkekic ?

mmkekic replied:
I just want to accept this PR so even a warning would be fine for me :p
But if you dont want to do sloppy job add a function that checks if the file is empty in this PR

jmalbos replied:
> I just want to accept this PR so even a warning would be fine for me :p

I'd honestly follow this approach. Print a warning for now and leave as a to-do the "empty-file" checks, which could even require modifying the existing tests. I would even volunteer to do that…

gonzaponte replied:
Fine, go ahead.

gonzaponte commented on invisible_cities/cities/components.py:
```diff
@@ -152,6 +154,39 @@ def print_every_loop(target):
+            try:
+                event_numbers_in_file = h5in.root.MC.extents.cols.evt_number[:]
+                event_numbers_to_copy = set(event_numbers_in_file).intersection(event_numbers)
+                mcinfo_io.copy_mc_info(h5in, writer, list(event_numbers_to_copy))
```
```suggestion
                mcinfo_io.copy_mc_info(h5in, writer, sorted(event_numbers_to_copy))
```
How aout this? I am afraid the set returns them in a random order

mmkekic replied:
But wait, this is wrong, we want to have them in the same order as in all the other tables, which means they cant be sorted but have the exact order of event_numbers list

gonzaponte replied:
You are right... I suggest
```
events_numbers_to_copy = tuple(evt for evt in event_numbers if evt in event_numbers_in_file)
```
or some other implementation using `filter`

jmalbos replied:
In fact, independently of the implementation, there is still a problem that I documented with a test. If the list `event_numbers` contains event numbers that are present in more than one file, then the output file will contain more events than `len(event_numbers)`. I'm not sure there is a way to fix that easily.

mmkekic replied:
I dont think you can solve the duplicated events in one file problem in this PR, so Id say forget about that possibility for now...And the efficiency of lists vs sets is really not important in this function, since the slowest part (by couple of orders of magnitude) is the writer itself

gonzaponte replied:
Yeah, we cannot deal with that right now

gonzaponte commented on invisible_cities/cities/components.py:
```diff
@@ -305,7 +332,7 @@ def hits_and_kdst_from_files(paths: List[str]) -> Iterator[Dict[str,Union[HitCol
                 event_number, timestamp = evtinfo.fetch_all_fields()
                 hits = hits_from_df(hits_df.loc[hits_df.event == event_number])
-                yield dict(hits = hits[event_number], kdst = kdst_df.loc[kdst_df.event==event_number], mc=mc_info, run_number=run_number,
+                yield dict(hits = hits[event_number], kdst = kdst_df.loc[kdst_df.event==event_number], run_number=run_number,
```
can we take the chance to split this line?

gonzaponte commented on invisible_cities/cities/components.py:
```diff
@@ -305,7 +332,7 @@ def hits_and_kdst_from_files(paths: List[str]) -> Iterator[Dict[str,Union[HitCol
-                yield dict(hits = hits[event_number], kdst = kdst_df.loc[kdst_df.event==event_number], mc=mc_info, run_number=run_number,
+                yield dict(hits = hits[event_number], kdst = kdst_df.loc[kdst_df.event==event_number], run_number=run_number,
                            event_number=event_number, timestamp=timestamp)
             # NB, the monte_carlo writer is different from the others:
```
You forgot to remove this comment that you removed from other readers.

gonzaponte commented on invisible_cities/cities/components_test.py:
```diff
@@ -160,3 +164,30 @@ def test_hits_and_kdst_from_files(ICDATADIR):
+    fl.push(source = the_source,
+            pipe   = fl.pipe(the_collector.sink))
+
+    assert the_source == the_collector.future.result()
```
You are not meant to call the `result` method yourself. Rather you should use the `result` argument of `push` to define your futures. Something like
```
result = fl.push(source = ..., pipe =..., result = the_collector.future)

assert the_source == result
```
This is (still) undocumented, but `result` (the variable) in this snippet can be different things depending on what you write in the `result` argument:
- If you write a single future, it `result` will be the value of that future
- If you write a tuple of futures, it will be a tuple with the value of each future
- If you write a mapping of futures, it will be a namespace with the value of each future

gonzaponte commented on invisible_cities/cities/components_test.py:
```diff
@@ -160,3 +164,30 @@ def test_hits_and_kdst_from_files(ICDATADIR):
+    assert the_source == the_collector.future.result()
+
+
+def test_copy_mc_info_noMC(ICDATADIR, config_tmpdir):
+    file_in  = os.path.join(ICDATADIR, 'run_2983.h5')
+    file_out = os.path.join(config_tmpdir, 'dummy_out.h5')
+    with tb.open_file(file_out, "w") as h5out:
+        copy_mc_info([file_in], h5out, [])
```
Is this meant to check that it doesn't fail without MC tables?
As I said before I think it should and the test should reflect that.

gonzaponte commented on invisible_cities/cities/components_test.py:
```diff
@@ -160,3 +164,30 @@ def test_hits_and_kdst_from_files(ICDATADIR):
+    with tb.open_file(file_out, 'w') as h5out:
+        copy_mc_info([file_in, file_in], h5out, [0,1,0,9])
+        events_in_h5out = h5out.root.MC.extents.cols.evt_number[:]
+        print(events_in_h5out)
```
Remove print

gonzaponte commented on invisible_cities/cities/diomira.py:
```diff
@@ -85,33 +83,37 @@ def diomira(files_in, file_out, compression, event_range, print_mod, detector_db
+                                          fl.branch(write_evt_filter)                   ,
+                                          trigger_filter.filter                         ,
+                                          simulate_sipm_response_                       ,
+                                          fl.branch("event_number", evtnum_collect.sink),
+                                          fl.fork(write_pmt                             ,
```
Why not have it as one more pipe in the fork?

gonzaponte commented on invisible_cities/cities/diomira.py:
```diff
@@ -85,33 +83,37 @@ def diomira(files_in, file_out, compression, event_range, print_mod, detector_db
+                                       evtnum_list   = evtnum_collect.future,
+                                       events_filter = trigger_filter.future))
+
+        copy_mc_info(files_in, h5out, result.evtnum_list) if run_number <= 0 else None
```
Mmm this syntax is a bit fishy... Don't use inline `if/else` for something that is not being collected anywhere.
Better
```
if run_number <= 0:
    copy_mc_info(...)
```

gonzaponte commented on invisible_cities/dataflow/dataflow.py:
```diff
@@ -239,6 +239,19 @@ def spy_count():
+    else:
+        raise ValueError("dataflow.collect_into_list only accepts one argument")
+
+
```
Is this a leftover from the discussion or are we keeping it for some reason?

gonzaponte replied:
Same for the tests affecting this

jmalbos replied:
Yes, I have yet to push the commits removing the code from `dataflow`.

mmkekic replied:
On a side note, since this PR will need some history cleaning, @gonzaponte  do you want to keep these changes in the history if they turn out to be useful in the future dataflow refinements or you prefer to remove them?

gonzaponte replied:
Good question... We can maybe leave them there, if that doesn't add too much noise.

gonzaponte commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -3,9 +3,12 @@
 import pandas as pd
 
+from copy import deepcopy
+
```
why this import? I guess is some leftover from different attempts

gonzaponte commented on invisible_cities/io/mcinfo_io.py:
```diff
@@ -168,6 +173,30 @@ def __call__(self, mctables: (tb.Table, tb.Table, tb.Table, tb.Table),
+        for n in which_events:
+            writer(mctables=mcinfo, evt_number=n)
+    except NoParticleInfoInFile:
+        pass
```
Again, I don't think this should pass silently. I think we want it to crash whenever there is a problem in the input data. That's why we have `if run_number <= 0 ` to deal with that case in the cities.

gonzaponte commented on invisible_cities/io/mcinfo_io_test.py:
```diff
@@ -176,6 +177,18 @@ def test_mc_info_writer_filter_first_event_of_first_file(output_tmpdir, ICDATADI
         assert all(x<y for x, y in zip(last_hit_list, last_hit_list[1:]))
 
 
+def test_copy_mc_info_which_events_is_none(ICDATADIR, config_tmpdir):
+    file_in  = os.path.join(ICDATADIR, "Kr83_nexus_v5_03_00_ACTIVE_7bar_10evts.sim.h5")
+    file_out = os.path.join(config_tmpdir, "dummy_out.h5")
+
+    with tb.open_file(file_out, 'w') as h5out:
+        writer = mc_info_writer(h5out)
+        with tb.open_file(file_in, 'r') as h5in:
+            copy_mc_info(h5in, writer)
+            events_in_h5in  = h5in .root.MC.extents.cols.evt_number[:]
+            events_in_h5out = h5out.root.MC.extents.cols.evt_number[:]
+            assert all(events_in_h5in == events_in_h5out)
+
```
This is good. Can we have another one with a subset of events?

jmalbos committed:
Remove dataflow.collect_into_list and tests

gonzaponte commented on invisible_cities/cities/components.py:
```diff
@@ -44,6 +44,8 @@
 from .. filters.s1s2_filter       import               pmap_filter
 from .. database                  import                   load_db
 from .. sierpe                    import                       blr
+from .. reco                      import             tbl_functions as tbl
```
Rmove unused import

gonzaponte reviewed (approved):
This PR applies a new paradigm to the propagation of the MC information through the cities. Instead of introducing it in the dataflow pipe, it is copied at the end, taking into account which events were filtered out. This will be useful in conjunction with upcoming PRs where the format of the MC data is modified.

The new functions introduced here are very modular and well tested. Depending on the evolution of the dataflow package, the implementation of the evt_id collector may change in the future, as indicated in the commit history.

Great first PR. Good job!
